rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,610 @@
+more important that upgrade systems be reliable than fast, and the simplest way to synchronize
+controllers is to have a single short-lived *global* controller lease.  A controller publishes
+what its working on, and how long it will be working.  Other controllers wait until after that
+lease either expires, or is explicitly released (preferable).  After release, only nodes who
+have heartbeated since the release time should be considered valid upgrade targets (so that the
+version information in the backend matches the post-upgrade state of the world).
+
+Regardless of the lease strategy, leases should include some basic information, including:
+
+- Name of the controller that created the lease.
+- Est. number of servers targeted for upgrade (this might be the number of servers which
+match the `server_selector` and not necessarily reflective of the number of servers that
+will actually get upgraded).
+- Version being targeted.
+
+
+#### Upgrade UX
+
+At the cluster-level, upgrade flows should roughly fall into two camps.  Manual and automatic,
+where manual upgrades require some kind of triggering from a user.  Other than needing a simple
+manual trigger, the UX around these two modes of operation should be identical.  I think the
+best way to do this is to have the process of applying policies to available releases produce
+a compiled "target state" that is independent of the resources used to calculate it (i.e. the
+upgrade policies and release manifests).
+
+By keeping the target state indipendent of the resources used to calculate it, we are able to
+store it as an indipendent resource.  Upgrade controllers will attempt to converge toward the stored
+target state, meaning that they will not be directly affected by new releases becoming available
+or upgrade policies being changed.  If the cluster is configured for automatic upgrades, then
+periodic operations in the auth server can recompile the target state.  If the cluster is configured
+for manual upgrades, the user can preview and apply new compilations of the target state via `tctl`.
+
+This should allow us to implement a manual upgrade flow that looks something like this:
+
+```
+$ tctl upgrade
+<diff of target state>
+Proceed: y/n
+```
+
+Or, if we want to allow for manual modification of target state, something like:
+
+```
+$ tctl upgrade plan > target-state.yaml
+# review the plan & make changes
+
+$ tctl upgrade apply -f target-state.yaml
+<diff>
+Proceed: y/n
+```
+
+### Technical Notes
+
+
+#### Crypto
+
+We aren't in full control of the cryptography at play since we're integrating user scripts and
+third-party installation mechanisms, but where possible we will be opinionated.  Where applicable
+we will select algorithms that are relatively modern and targeted to meet or exceede a 128-bit
+security level.
+
+##### Package Signing
+
+This is generally covered by existing package distribution solutions.  I'm leaning toward `minisign`
```

Comment:
So one of the things to watch out for with minisign is I don't believe it's compatible with fips validated crypto. When I looked at this for gravity, the thought was that we would need to port minisign to support signatures covered by FIPS validated crypto if it came up.

Reply:
I would check if FedRamp actually require signatures on packages at all. For example, you can download the teleport docker container with no signatures at all and run it as fips mode no problem. If this is the case, can we get away with just using minisign for all cases :) If FedRamp does not concern itself with the crypto used for package signatures, we have escaped this conversation :)

Reply:
AFAIK fedramp doesn't work like that, as I remember from last year, it's sort of based around if you're not using fips validated crypto it's considered you have no crypto. And while not covered by specific controls, i think it had something like if there is a security design that requires crypto and it's not fips it doesn't count. It's been awhile since I read it.

I think Fedramp or nist is doing a review right now on supply chains as well and have been doing lots of industry reviews (i've posted some of the drafts in slack over the last year). So some of these recommendations I think do include docker content trust... not sure off hand if that can be made fips compliant though, but I thought docker had a fips build.

Reply:
So to forrest's point, fips can be out of scope and not have fleet upgrades, wait for edwards curve to get approved and someone takes edwards through the CMP program to have an approved module. 

Or fork minisign and use P-258 signatures or the like, or go the gpg route / hook into the package manager gpg security already in use.

Reply:
I haven't come across guidance that was specific about package signing (though there plenty of general guidelines about needing to know the origin of software, and needing to ensure integrity).  Most of what I've come across *could* be interpreted to just mean "use TLS".  Why risk it though?  I think its reasonable to just require that upgrades in FIPS mode require a fips compatible package manger be available for the time being.

If we go with using github as our "index server" we'll also need to provide a fips compliant GPG signature for the index file.  Once again, not what we want the default behavior to be, but if we make the index file verification behavior configurable, then folks can choose to have it depend on a local FIPS compliant GPG install (only necessary for auth servers).